lib/pbench/server/api/resources/upload_api.py

```diff
@@ -355,27 +354,95 @@ def put(self, filename: str):
                         f"Failed to write .md5 file '{md5_full_path}'",
                     )
 
-                # Create a file tree object
-                try:
-                    file_tree = FileTree(self.config, self.logger)
-                    self.logger.info(
-                        "Created {} of type {}", file_tree, type(file_tree)
-                    )
-                except Exception:
-                    raise CleanupTime(
-                        HTTPStatus.INTERNAL_SERVER_ERROR, "Unable to map the file tree"
-                    )
+            # Create a file tree object
+            try:
+                file_tree = FileTree(self.config, self.logger)
+            except Exception:
+                raise CleanupTime(
+                    HTTPStatus.INTERNAL_SERVER_ERROR, "Unable to map the file tree"
+                )
 
-                # Move the files to their final location
-                try:
-                    tarball = file_tree.create(controller, tar_full_path)
-                except Exception:
-                    raise CleanupTime(
-                        HTTPStatus.INTERNAL_SERVER_ERROR,
-                        "Unable to create dataset in file system",
+            # Move the files to their final location
+            try:
+                tarball = file_tree.create(controller, tar_full_path)
+            except Exception:
+                raise CleanupTime(
+                    HTTPStatus.INTERNAL_SERVER_ERROR,
+                    f"Unable to create dataset in file system for {tar_full_path}",
+                )
+
+            # From this point, failure will remove the tarball from the file
+            # tree.
+            #
+            # TODO: the Tarball.delete method won't clean up empty controller
+            # directories, but to call FileTree.delete(dataset.name) we'd need
+            # to add parameters to the recovery object.
```

Comment:
Perhaps note only that `tarball.delete` won't clean up empty controller directories, and create a tracking issue for the rest?